src/datarobotx/idp/custom_application_source_version.py

```diff
@@ -0,0 +1,256 @@
+                    method="POST",
+                    url=f"customApplicationSources/{custom_application_source_id}/versions/",
+                    data=encoder,
+                    headers=headers,
+                )
+            new_version = response.json()["id"]
+
+        return str(new_version)
+
+
+# Public functions
+def get_or_create_custom_application_source_version(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    **kwargs: Any,
+) -> str:
+    """Get or create a custom application source version.
+
+    Parameters
+    ----------
+    custom_application_source_id : str
+        The ID of the custom application source.
+    **kwargs : Any
+        Additional keyword arguments.
+
+    Returns
+    -------
+    str
+        The ID of the custom application source version.
+    """
+    runtime_parameter_values = kwargs.pop("runtime_parameter_values", None)
+
+    app_source_version_id = _get_or_create_custom_application_source_version(
+        endpoint=endpoint,
+        token=token,
+        custom_application_source_id=custom_application_source_id,
+        runtime_parameter_values=runtime_parameter_values,
+        **kwargs,
+    )
+    if runtime_parameter_values is not None:
+        app_source_version_id = _get_or_create_custom_application_source_version(
+            endpoint=endpoint,
+            token=token,
+            custom_application_source_id=custom_application_source_id,
+            previous_custom_application_source_version_id=app_source_version_id,
+            runtime_parameter_values=runtime_parameter_values,
+            **kwargs,
+        )
+    return app_source_version_id
+
+
+def get_or_create_custom_application_source_version_from_previous(
+    endpoint: str,
+    token: str,
+    custom_application_source_id: str,
+    runtime_parameter_values: Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]] = None,
+    **kwargs: Any,
+) -> str:
+    """Get or create a custom application source version from a previous version.
+
+    Parameters
+    ----------
+    custom_application_source_id : str
+        The ID of the custom application source.
+    runtime_parameter_values : Optional[List[Union[RuntimeParameterValue, Dict[str, str]]]], optional
+        A list of runtime parameter values. Each value can be either a `RuntimeParameterValue` object or a dictionary
+        with parameter names as keys and parameter values as values. (default: None)
+    **kwargs : Any
+        Additional keyword arguments.
+
+    Returns
+    -------
+    str
+        The ID of the custom application source version.
+    """
+    if runtime_parameter_values is not None:
+        runtime_parameter_values_objs = [
+            RuntimeParameterValue(**d) for d in runtime_parameter_values if isinstance(d, dict)
+        ]
+    else:
+        runtime_parameter_values_objs = None
+
+    client = dr.Client(token=token, endpoint=endpoint)  # type: ignore
+    app_version_token = get_hash(
```

Comment:
i think the previous verison id lookup needs to happen before this and the id should be included in the hash for this to be idempotent; also we should raise an error if this lookup fails (e.g. no previous versions at all; this is organized a bit different from the CMV helper and i haven't thought deeply about if that's concerning, but am just noting it now for discussion as needed

Reply:
```
except:
    raise ValueError("No previous version found")
```
in line 254 is raising the error if there are no previous versions

I am using the hash in the call to find the previous version - so I need to calculate the token before, no?

The whole thing got slightly confusing with several refactors. I was relying on the tests to convince myself of the correct behaviour

Reply:
I refactored out one function to make it a bit more clear what's happening.